pkg/controllers/routereflector/topologies/multi.go

```diff
@@ -0,0 +1,318 @@
+	zones := []string{}
+	rrPerZone := map[string][]*corev1.Node{}
+
+	// Create per zone RR lists for MZR selection
+	if t.Config.ZoneLabel != "" {
+		for i, rr := range routeReflectors {
+			rrZone := rr.GetLabels()[t.Config.ZoneLabel]
+			log.Debugf("RR:%s's zone: %s", rr.GetName(), rrZone)
+			rrPerZone[rrZone] = append(rrPerZone[rrZone], routeReflectors[i])
+		}
+
+		for zone := range rrPerZone {
+			zones = append(zones, zone)
+		}
+		sort.Strings(zones)
+	}
+
+	// Let's find RRs for each nodes
+	for _, n := range nodeList {
+		if t.IsRouteReflector(string(n.GetUID()), n.GetLabels()) {
+			continue
+		}
+
+		routeReflectorsForNode := []*corev1.Node{}
+
+		// Run MZR selection first for nodes if zone label is configured
+		if t.Config.ZoneLabel != "" {
+			nodeZone := n.GetLabels()[t.Config.ZoneLabel]
+			log.Debugf("Node's zone: %s", nodeZone)
+
+			// Select the 1st RR from the same zone if there're any
+			if len(rrPerZone[nodeZone]) > 0 {
+				rr := selectRRfromZone(rrIndexPerZone, rrPerZone, nodeZone)
+				log.Debugf("Adding %s as 1st RR for Node:%s", rr.GetName(), n.GetName())
+				routeReflectorsForNode = append(routeReflectorsForNode, rr)
+			}
+
+			// Select the 2nd RR from a different zone
+			for _, zone := range zones {
+				if zone != nodeZone {
+					rr := selectRRfromZone(rrIndexPerZone, rrPerZone, zone)
+					log.Debugf("Adding %s as 2nd RR for Node:%s", rr.GetName(), n.GetName())
+					routeReflectorsForNode = append(routeReflectorsForNode, rr)
+					break
+				}
+			}
+		}
+
+		// Calcualte missing RR peers for node
+		peers := int(math.Min(float64(len(routeReflectors)), float64(t.Config.ReflectorsPerNode)))
+
+		// Selecting the remaning RRs sequentially with round robin
+		for len(routeReflectorsForNode) < peers {
+			rrIndex++
+			if rrIndex == len(routeReflectors) {
+				// Jump back to first
+				rrIndex = 0
+			}
+
+			rr := routeReflectors[rrIndex]
+			if isAlreadySelected(routeReflectorsForNode, rr) {
+				continue
+			}
+
+			log.Debugf("Adding %s to RRs of %s", rr.GetName(), n.GetName())
+			routeReflectorsForNode = append(routeReflectorsForNode, rr)
+		}
+
+		// Node selector template
+		nodeSelector := fmt.Sprintf("%s=='%s'", t.Config.HostnameLabel, n.GetLabels()[t.Config.HostnameLabel])
+
+		// Generate or update BGP peer config for node
+		for _, rr := range routeReflectorsForNode {
+			rrID := getRouteReflectorID(string(rr.GetUID()))
+			name := fmt.Sprintf(DefaultRouteReflectorClientName+"-%s", rrID, n.GetUID())
+
+			clientConfig := findBGPPeer(existingPeers, name)
+			if clientConfig != nil {
+				// The rule is already exists, content check not necessary because they are unique
+				toKeep[clientConfig.GetName()] = true
+				continue
+			}
+
+			// Create new one
+			log.Debugf("New BGPPeers: %s", name)
+			clientConfig = generateBGPPeerStub(name)
+			clientConfig.Spec = apiv3.BGPPeerSpec{
+				NodeSelector: nodeSelector,
+				PeerSelector: fmt.Sprintf("%s=='%d'", t.NodeLabelKey, rrID),
```

Comment:
A couple options that might be nicer than creating a BGPPeer object per RR to peer with.

- Use the `in` selector - e.g., `hostname in {"node1", "node2", "node3"`
- Apply a label to the RRs, and use a static selector - e.g., `peerSelector: has(my-hostname)` - control peering by adding / removing labels from the RRs themselves. 

I don't think there is fundamentally a problem with multiple BGP peer objects, except that the number of peers might get quite large depending on the number of nodes and RRs. Just thought I'd suggest some alternatives in case you hadn't considered them.